In [1]:
# importing the required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt
%matplotlib inline

In [2]:
df = pd.read_csv('Mart.csv')

In [3]:
df.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106

In [6]:
# to check if there is any null values in the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6435 entries, 0 to 6434
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Store         6435 non-null   int64  
 1   Date          6435 non-null   object 
 2   Weekly_Sales  6435 non-null   float64
 3   Holiday_Flag  6435 non-null   int64  
 4   Temperature   6435 non-null   float64
 5   Fuel_Price    6435 non-null   float64
 6   CPI           6435 non-null   float64
 7   Unemployment  6435 non-null   float64
dtypes: float64(5), int64(2), object(1)
memory usage: 402.3+ KB


## 1. Which store has maximum sales

In [7]:
s = df[df["Store"] == 1]["Weekly_Sales"].sum()

In [8]:
sales = {}
for i in range(1,46):
    sales[i] = df[df["Store"] == i]["Weekly_Sales"].sum()

In [9]:
sales

{1: 222402808.85,
 2: 275382440.97999996,
 3: 57586735.06999999,
 4: 299543953.38,
 5: 45475688.900000006,
 6: 223756130.64000002,
 7: 81598275.14,
 8: 129951181.12999998,
 9: 77789218.99000001,
 10: 271617713.89,
 11: 193962786.8,
 12: 144287230.15,
 13: 286517703.8,
 14: 288999911.34000003,
 15: 89133683.92,
 16: 74252425.39999999,
 17: 127782138.83000003,
 18: 155114734.21000004,
 19: 206634862.09999996,
 20: 301397792.46000004,
 21: 108117878.91999999,
 22: 147075648.57,
 23: 198750617.85000002,
 24: 194016021.28000003,
 25: 101061179.16999999,
 26: 143416393.79,
 27: 253855916.88,
 28: 189263680.57999998,
 29: 77141554.30999999,
 30: 62716885.120000005,
 31: 199613905.5,
 32: 166819246.16000003,
 33: 37160221.95999999,
 34: 138249763.0,
 35: 131520672.08,
 36: 53412214.97,
 37: 74202740.32,
 38: 55159626.42,
 39: 207445542.46999997,
 40: 137870309.79,
 41: 181341934.89,
 42: 79565752.42999999,
 43: 90565435.41,
 44: 43293087.84,
 45: 112395341.42000002}

In [10]:
mx = 0
store = 0
for k in sales.items():
    if(k[1] > mx):
        mx = k[1]
        store = k[0]
print(store)

20


### so store 20 has maximum sales

## 2. Which store has maximum standard deviation i.e., the sales vary a lot. Also, find out the coefficient of mean to standard deviation (coefficient of variation)

In [11]:
# total number of stores
df["Store"].nunique()

45

In [13]:
# calculating the coefficient of variation and store having maximum and minimum standard deviation
mx_std_dev = 0;
mn_std_dev = 1000000000;
mx_store = 0
mn_store = 0
coeff = 0
print("The Coefficient of variations for stores are as follows :")
for i in range(1,46):
    std = df[df["Store"] == i].std()[1]
    mean = df[df["Store"] == i].mean()[1]
    if(std > mx_std_dev):
        mx_std_dev = std
        mx_store = i
    if(std < mn_std_dev):
        mn_std_dev = std
        mn_store = i
    coeff = (std/mean) *100
    print(i,coeff)
print("The store having maximum standard deviation is :",mx_store,"With standard deviation",mx_std_dev)
print("The store having minimum standard deviation is :",mn_store,"With standard deviation",mn_std_dev)

The Coefficient of variations for stores are as follows :
1 10.02921226813075
2 12.34238763319183
3 11.502140735338804
4 12.708253937002839
5 11.866844091939742
6 13.582285902663829
7 19.730468653665728
8 11.695283213906869
9 12.689546764678369
10 15.913349066639102
11 12.22618336004468
12 13.792532199977735
13 13.251362792719142
14 15.713673600948338
15 19.33839877822955
16 16.51806549478143
17 12.552067141232682
18 16.28454974226666
19 13.26801153482639
20 13.090268561738489
21 17.029239217349453
22 15.678287578148591
23 17.9721149197535
24 12.363737661375541
25 15.98604024072346
26 11.011066299216315
27 13.515544496695316
28 13.73297416512492
29 18.374246746816343
30 5.2008038555445175
31 9.016105262945947
32 11.831049176166283
33 9.286835290692942
34 10.82252383219858
35 22.968111389976414
36 16.25789124577327
37 4.208411895180792
38 11.087544692154392
39 14.990779108026006
40 12.342978096269446
41 14.817711243560769
42 9.03353280932982
43 6.4103629270192375
44 8.179331054568062
45

### The store having maximum standard deviation is : 14 With standard deviation 317569.9494755083
### The store having minimum standard deviation is : 37 With standard deviation 21837.4611900489

## 3. Some holidays have a negative impact on sales. Find out holidays which have higher sales than the mean sales in non-holiday season for all stores together

In [14]:
# mean sales of non holiday season for all stores together
mean_non_holiday = df[df["Holiday_Flag"] == 0].mean()[1]

In [15]:
mean_non_holiday

1041256.3802088555

In [18]:
# calculating the sales of different holiday seasons which have sales more than the mean sales of non holiday season
l = ["12-02-2010","11-02-2011","10-02-2012","10-09-2010","09-09-2011","07-09-2012","26-11-2010","31-12-2010","30-12-2011"]
for i in l:
    sales_holiday_stores = df[df["Date"] == i]["Weekly_Sales"].mean()
    if(sales_holiday_stores > mean_non_holiday):
        print("holiday Season : ",i,"||","Sales : ",sales_holiday_stores)

holiday Season :  12-02-2010 || Sales :  1074148.3917777783
holiday Season :  11-02-2011 || Sales :  1051915.3953333334
holiday Season :  10-02-2012 || Sales :  1111320.176
holiday Season :  07-09-2012 || Sales :  1074001.318
holiday Season :  26-11-2010 || Sales :  1462688.960888889


## 4. Provide a monthly and 6 month view of sales in units and give insights

In [21]:
# adding month and year column
df['Month'] = df["Date"].apply(lambda x : x.split('-')[1])
df['Year'] = df["Date"].apply(lambda x : x.split('-')[2])

In [22]:
df.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   

          CPI  Unemployment Month  Year  
0  211.096358         8.106    02  2010  
1  211.242170         8.106    02  2010  
2  211.289143         8.106    02  2010  
3  211.319643         8.106    02  2010  
4  211.350143         8.106    03  2010

In [23]:
# making year list
year_list = df["Year"].unique().tolist()

In [24]:
year_list

['2010', '2011', '2012']

In [28]:
# making monthly list
month_list = df["Month"].unique().tolist()

In [29]:
month_list

['02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12', '01']

In [30]:
# calculating the monthly sales
monthly_sales = {}
for i in year_list:
    for j in month_list:
        monthly_sales[j+"-"+i] = df[(df["Month"] == j) & (df["Year"] == i)]["Weekly_Sales"].sum()

### tabular data for monthly sales

In [31]:
monthly_sales

{'02-2010': 190332983.04000002,
 '03-2010': 181919802.5,
 '04-2010': 231412368.05,
 '05-2010': 186710934.34000003,
 '06-2010': 192246172.36,
 '07-2010': 232580125.98,
 '08-2010': 187640110.89,
 '09-2010': 177267896.37,
 '10-2010': 217161824.02,
 '11-2010': 202853370.14,
 '12-2010': 288760532.72,
 '01-2010': 0.0,
 '02-2011': 186331327.87,
 '03-2011': 179356448.29000002,
 '04-2011': 226526510.97,
 '05-2011': 181648158.16,
 '06-2011': 189773385.19,
 '07-2011': 229911398.87,
 '08-2011': 188599332.25,
 '09-2011': 220847738.42000002,
 '10-2011': 183261283.14999998,
 '11-2011': 210162354.87,
 '12-2011': 288078102.48,
 '01-2011': 163703966.82999998,
 '02-2012': 192063579.54000002,
 '03-2012': 231509650.49,
 '04-2012': 188920905.95,
 '05-2012': 188766479.45,
 '06-2012': 240610329.28999996,
 '07-2012': 187509452.39999998,
 '08-2012': 236850765.68,
 '09-2012': 180645544.47,
 '10-2012': 184361680.42000002,
 '11-2012': 0.0,
 '12-2012': 0.0,
 '01-2012': 168894471.66}

### Insight : the maximum sales in both year 2010 and  2011 are found in the month of december..may be because these are the closing dates and stores provide the maximum sales in these months. Moreover due to christmas and new year people purchase more goods.

In [32]:
# calculating half yearly sales. Due to the data constraints the last value if from August 2012 to October 2012 only i.e 3 months
half_yearly_sales = {}
half_yearly_sales["Feb-2010 to July-2010"] = monthly_sales["02-2010"]+monthly_sales["03-2010"]+monthly_sales["04-2010"]+monthly_sales["05-2010"]+monthly_sales["06-2010"]+monthly_sales["07-2010"]
half_yearly_sales["Aug-2010 to Jan-2011"] = monthly_sales["08-2010"]+monthly_sales["09-2010"]+monthly_sales["10-2010"]+monthly_sales["11-2010"]+monthly_sales["12-2010"]+monthly_sales["01-2011"]
half_yearly_sales["Feb-2011 to July-2011"] = monthly_sales["02-2011"]+monthly_sales["03-2011"]+monthly_sales["04-2011"]+monthly_sales["05-2011"]+monthly_sales["06-2011"]+monthly_sales["07-2011"]
half_yearly_sales["Aug-2011 to jan-2012"] = monthly_sales["08-2011"]+monthly_sales["09-2011"]+monthly_sales["10-2011"]+monthly_sales["11-2011"]+monthly_sales["12-2011"]+monthly_sales["01-2012"]
half_yearly_sales["Feb-2012 to July-2012"] = monthly_sales["02-2012"]+monthly_sales["03-2012"]+monthly_sales["04-2012"]+monthly_sales["05-2012"]+monthly_sales["06-2012"]+monthly_sales["07-2012"]
half_yearly_sales["Aug-2012 to Oct-2012"] = monthly_sales["08-2012"]+monthly_sales["09-2012"]+monthly_sales["10-2012"]

In [33]:
half_yearly_sales

{'Feb-2010 to July-2010': 1215202386.27,
 'Aug-2010 to Jan-2011': 1237387700.97,
 'Feb-2011 to July-2011': 1193547229.35,
 'Aug-2011 to jan-2012': 1259843282.8300002,
 'Feb-2012 to July-2012': 1229380397.12,
 'Aug-2012 to Oct-2012': 601857990.5699999}

# Statistical Model

## For store 2 develop Linear Regression model– Utilize variables like date and restructure dates as 1 for 5 Feb 2010 (starting from the earliest date in order). Hypothesize if CPI, unemployment, and fuel price have any impact on sales. 

In [36]:
# making the dataframe that contains entries of store 2 only
store2 = df[df["Store"] == 2]

In [37]:
store2.head()

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
143      2  05-02-2010    2136989.46             0        40.19       2.572   
144      2  12-02-2010    2137809.50             1        38.49       2.548   
145      2  19-02-2010    2124451.54             0        39.69       2.514   
146      2  26-02-2010    1865097.27             0        46.10       2.561   
147      2  05-03-2010    1991013.13             0        47.17       2.625   

            CPI  Unemployment Month  Year  
143  210.752605         8.324    02  2010  
144  210.897994         8.324    02  2010  
145  210.945160         8.324    02  2010  
146  210.975957         8.324    02  2010  
147  211.006754         8.324    03  2010

In [39]:
# importing the required libraries
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [41]:
# reshape using numpy.
y = store2["Weekly_Sales"].values.reshape(-1,1)

In [42]:
print(y)

[[2136989.46]
 [2137809.5 ]
 [2124451.54]
 [1865097.27]
 [1991013.13]
 [1990483.78]
 [1946070.88]
 [1750197.81]
 [2066187.72]
 [1954689.21]
 [1874957.94]
 [1821990.93]
 [1802450.29]
 [2042581.71]
 [1880752.36]
 [1896937.1 ]
 [1957113.89]
 [2102539.93]
 [2025538.76]
 [2001636.96]
 [1939927.09]
 [2003940.64]
 [1880902.62]
 [1845879.79]
 [1781717.71]
 [1804246.16]
 [1991909.98]
 [1895601.05]
 [1964335.23]
 [1863840.49]
 [1904608.09]
 [1839128.83]
 [1793903.6 ]
 [1724557.22]
 [1827440.43]
 [1849921.44]
 [1794355.49]
 [1737947.64]
 [1802755.11]
 [1939061.41]
 [1916812.74]
 [1956739.17]
 [2658725.29]
 [2015781.27]
 [2378726.55]
 [2609166.75]
 [3436007.68]
 [1750434.55]
 [1758050.79]
 [1744193.58]
 [1751384.9 ]
 [1695371.68]
 [1929346.23]
 [2168041.61]
 [2080884.82]
 [1833511.08]
 [1981607.78]
 [1879107.31]
 [1902557.66]
 [1766162.05]
 [1800171.36]
 [1847552.61]
 [1856467.84]
 [1886339.6 ]
 [1745545.28]
 [1837743.6 ]
 [1838513.07]
 [1688281.86]
 [1797732.56]
 [1933756.21]
 [1929153.16]
 [1953

In [43]:
# features
X = store2.iloc[:,3:8].values

In [44]:
print(X)

[[  0.         40.19        2.572     210.7526053   8.324    ]
 [  1.         38.49        2.548     210.8979935   8.324    ]
 [  0.         39.69        2.514     210.9451605   8.324    ]
 [  0.         46.1         2.561     210.9759573   8.324    ]
 [  0.         47.17        2.625     211.0067542   8.324    ]
 [  0.         57.56        2.667     211.037551    8.324    ]
 [  0.         54.52        2.72      210.8733316   8.324    ]
 [  0.         51.26        2.732     210.6766095   8.324    ]
 [  0.         63.27        2.719     210.4798874   8.2      ]
 [  0.         65.41        2.77      210.2831653   8.2      ]
 [  0.         68.07        2.808     210.1495463   8.2      ]
 [  0.         65.11        2.795     210.1000648   8.2      ]
 [  0.         66.98        2.78      210.0505833   8.2      ]
 [  0.         71.28        2.835     210.0011018   8.2      ]
 [  0.         73.31        2.854     209.9984585   8.2      ]
 [  0.         74.83        2.826     210.2768443   8.2

In [46]:
# making the object of class implementing Linear Regression 
model = LinearRegression()

In [47]:
# splitting the dataset into 80-20 format
XTrain,XTest,YTrain,YTest = train_test_split(X,y,test_size = 0.2,random_state = 0)

In [48]:
print(XTrain.shape)
print(YTrain.shape)

(114, 5)
(114, 1)


In [50]:
# training (i.e fitting the line)
model.fit(XTrain,YTrain)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [51]:
# prediction over the unseen testing data
y_prediction = model.predict(XTest)
print(y_prediction)

[[1946426.32066277]
 [1879951.40417502]
 [1882260.89155458]
 [1909126.59112045]
 [1892200.14285578]
 [1997585.75845128]
 [1927578.52051966]
 [1813000.5457226 ]
 [1833659.49252555]
 [1999323.51369375]
 [2071612.75963028]
 [1896266.45746001]
 [1928436.96804825]
 [2092813.80122205]
 [1818429.20917997]
 [1925316.57926143]
 [1871939.28466764]
 [1904079.49057578]
 [1901651.0150834 ]
 [1896505.41482992]
 [1811046.25035841]
 [1953992.22854554]
 [1881372.17767078]
 [1902070.30684743]
 [1818201.52222804]
 [1930985.06653398]
 [1824031.3024909 ]
 [1969032.74561768]
 [1961000.05765377]]


In [52]:
# coefficients of different features in the hypothesis
print(model.coef_)

[[  58401.325891     -2238.23488176 -257714.5342688    54515.96326202
   281398.25517897]]


### The coefficients that are positive signifies that, as the value of the feature corresponding to this coefficient increases the value of dependent variable increases (i.e. sales in this case and vice versa.
### Similarly high value of the coefficient means the feature corresponding to this coefficient is more related to the dependent variable( i.e. sales in this case) and vice versa.

#### Therefore Holiday flag have positive coefficient with value 58401.325
#### Therefore Temperature have negative coefficient with value 2238.234
#### Therefore Fuel Price have negative coefficient with value 257714.534
#### Therefore CPI have positive coefficient with value 54515.963
#### Therefore Unemployment have positive coefficient with value 281398.255

# THANK YOU